src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.Android.cs

```diff
@@ -228,22 +233,110 @@ sealed class ModalContainer : ViewGroup
 					.Commit();
+			}
 
+			protected override void OnAttachedToWindow()
+			{
+				base.OnAttachedToWindow();
 				UpdateMargin();
+				UpdateRootView(GetWindowRootView());
+			}
+
+			protected override void OnDetachedFromWindow()
+			{
+				base.OnDetachedFromWindow();
+				UpdateRootView(null);
+			}
+
+			void UpdateRootView(AView? rootView)
+			{
+				if (_rootView.IsAlive() && _rootView != null)
+				{
+					_rootView.LayoutChange -= OnRootViewLayoutChanged;
+					_rootView = null;
+				}
+
+				if (rootView.IsAlive() && rootView != null)
+				{
+					rootView.LayoutChange += OnRootViewLayoutChanged;
+					_rootView = rootView;
+					_currentRootViewHeight = _rootView.MeasuredHeight;
+					_currentRootViewWidth = _rootView.MeasuredWidth;
+				}
+			}
+
+			// If the RootView changes sizes that means we also need to change sizes
+			// This will typically happen when the user is opening the soft keyboard 
+			// which sometimes causes the available window size to change
+			void OnRootViewLayoutChanged(object? sender, LayoutChangeEventArgs e)
+			{
+				if (Modal == null || sender is not AView view)
+					return;
+
+				var modalStack = Modal?.Navigation?.ModalStack;
+				if (modalStack == null ||
+					modalStack.Count == 0 ||
+					modalStack[modalStack.Count - 1] != Modal)
+				{
+					return;
+				}
+
+				if ((_currentRootViewHeight != view.MeasuredHeight || _currentRootViewWidth != view.MeasuredWidth)
+					&& this.ViewTreeObserver != null)
+				{
+					// When the keyboard closes Android calls layout but doesn't call remeasure.
+					// MY guess is that this is due to the modal not being part of the FitSystemWindowView
+					// The modal is added to the decor view so its dimensions don't get updated.
+					// So, here we are waiting for the layout pass to finish and then we remeasure the modal					
+					//
+					// For .NET 8 we'll convert this all over to using a DialogFragment
+					// which means we can delete most of the awkward code here
+					_currentRootViewHeight = view.MeasuredHeight;
+					_currentRootViewWidth = view.MeasuredWidth;
+					if (!this.IsInLayout)
+					{
+						this.InvalidateMeasure(Modal);
+						return;
+					}
+
+					_rootViewLayoutListener ??= new GenericGlobalLayoutListener((listener, view) =>
+					{
+						if (view != null && !this.IsInLayout)
+						{
+							listener.Invalidate();
+							_rootViewLayoutListener = null;
+							this.InvalidateMeasure(Modal);
+						}
+					}, this);
+				}
 			}
 
 			void UpdateMargin()
 			{
 				// This sets up the modal container to be offset from the top of window the same
 				// amount as the view it's covering. This will make it so the
-				// ModalContainer takes into account the statusbar or lack thereof
-				var rootView = GetWindowRootView();
-				int y = (int)rootView.GetLocationOnScreenPx().Y;
+				// ModalContainer takes into account the StatusBar or lack thereof
+				var decorView = Context?.GetActivity()?.Window?.DecorView;
```

Comment:
yea, the previous code was using the `WindowRootView` to try and figure out the offsets on the modal page that it needed to use so that the `Modal` view is placed correctly below the status bar and above any of the device action buttons. That worked mostly fine, except when we switched to `AdjustPan` which causes `Android` to pan the `WindowRootView`, at which point I can't really use its location to figure out the modal offsets. 

I found better APIs (below this code) that seem to give the correct values I need for the `modal` without trying to use the root window view. I tested on an API 21 device, a physical pixel device, and a few different emulators and it seems to all work fine. 

With `Xamarin.Forms` we didn't do much of this because in `Xamarin.Forms` we had a root `ViewGroup` and then the primary view was added to that (very similar to how WinUI currently works). With MAUI we got rid of that `ViewGroup` so I didn't really have anywhere to add the `Modal` to. I decided to just try adding to the `DecorView` but that has proved much more difficult than I'd hoped. Android does a lot of "magic" that the modal is no longer opted into. 

For .NET 8 I'm hoping to just replace all the `modal` code with `DialogFragment`. I have that code mostly working but that seems like too grand of a change to back port so this PR should get it all working well enough until .NET 8.